crypto/asn1/d2i_param.c

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include "internal/cryptlib.h"
+#include <openssl/evp.h>
+#include <openssl/asn1.h>
+#include "internal/evp_int.h"
+#include "internal/asn1_int.h"
+
+EVP_PKEY *d2i_KeyParams(int type, EVP_PKEY **a, const unsigned char **pp,
+                        long length)
+{
+    EVP_PKEY *ret;
+    const unsigned char *p = *pp;
+
+    if ((a == NULL) || (*a == NULL)) {
+        if ((ret = EVP_PKEY_new()) == NULL) {
+            ASN1err(ASN1_F_D2I_KEYPARAMS, ERR_R_EVP_LIB);
+            return NULL;
+        }
+    } else
+        ret = *a;
+
+    if (type != EVP_PKEY_id(ret) && !EVP_PKEY_set_type(ret, type)) {
+        ASN1err(ASN1_F_D2I_KEYPARAMS, ASN1_R_UNSUPPORTED_TYPE);
+        goto err;
+    }
+
+    if (ret->ameth->param_decode == NULL
+        || !ret->ameth->param_decode(ret, &p, length)) {
+            ASN1err(ASN1_F_D2I_KEYPARAMS, ERR_R_ASN1_LIB);
```

Comment:
I haven't looked how i2d_pr.c does this, but it seems to me that `ret->ameth->parqm_decode == NULL` should generate a `ASN1_R_UNSUPPORTED_TYPE` error, i.e that check should be in the previous if statement...

Reply:
Furthermore, there's no real need to add another error when `param_decode` fails...

Reply:
BTW, you should probably check that `ret->ameth` isn't NULL

Reply:
hmm well in that case none of the errors should be added - since the lower levels add errors for these branch paths.

Reply:
> hmm well in that case none of the errors should be added - since the lower levels add errors for these branch paths.

This is a matter of judgement, really.  What the `ERR_R_ASN1_LIB` error does is to give a sort of traceback, showing in the error printout that the code went through here.  I'm mostly thinking that since the i2d code does *not* add this kind of error, it would be consistent if this didn't either.